DataFormats/Portable/interface/AlpakaHostCollection.h

```diff
@@ -0,0 +1,80 @@
+#define DataFormats_Portable_interface_AlpakaHostCollection_h
+
+#include <optional>
+
+#include <alpaka/alpaka.hpp>
+
+#include "DataFormats/Portable/interface/AlpakaCollection.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/alpaka/config.h"
+
+// generic SoA-based product in pinned host memory
+template <typename T>
+class AlpakaCollection<T, alpaka_common::DevHost> {
+public:
+  using Buffer = alpaka::Buf<alpaka_common::DevHost, std::byte, alpaka::DimInt<1u>, uint32_t>;
+
+  AlpakaCollection() : buffer_{}, layout_{} {
+#ifdef DEBUG_COLLECTION_CTOR_DTOR
+    std::cout << "[partial template specialisation]" << std::endl;
+    std::cout << __PRETTY_FUNCTION__ << " [this=" << this << "]" << std::endl;
+#endif  // DEBUG_COLLECTION_CTOR_DTOR
+  }
+
+  AlpakaCollection(int32_t elements, alpaka_common::DevHost const &host)
+      : buffer_{alpaka::allocBuf<std::byte, uint32_t>(
+            host, alpaka::Vec<alpaka::DimInt<1u>, uint32_t>{T::compute_size(elements)})},
+        layout_{elements, buffer_->data()} {
+    // Alpaka set to a default alignment of 128 bytes defining ALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128
+    assert(reinterpret_cast<uintptr_t>(buffer_->data()) % T::alignment == 0);
+#ifdef DEBUG_COLLECTION_CTOR_DTOR
+    std::cout << "[partial template specialisation]" << std::endl;
+    std::cout << __PRETTY_FUNCTION__ << " [this=" << this << "]" << std::endl;
+#endif  // DEBUG_COLLECTION_CTOR_DTOR
+  }
+
+  template <typename TDev>
+  AlpakaCollection(int32_t elements, alpaka_common::DevHost const &host, TDev const &device)
+      : buffer_{alpaka::allocMappedBuf<std::byte, uint32_t>(
+            host, device, alpaka::Vec<alpaka::DimInt<1u>, uint32_t>{T::compute_size(elements)})},
+        layout_{elements, buffer_->data()} {
+    // Alpaka set to a default alignment of 128 bytes defining ALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128
+    assert(reinterpret_cast<uintptr_t>(buffer_->data()) % T::alignment == 0);
+  }
+
+  ~AlpakaCollection() {
+#ifdef DEBUG_COLLECTION_CTOR_DTOR
+    std::cout << "[partial template specialisation]" << std::endl;
+    std::cout << __PRETTY_FUNCTION__ << " [this=" << this << "]" << std::endl;
+#endif  // DEBUG_COLLECTION_CTOR_DTOR
+  }
+
+  // non-copyable
+  AlpakaCollection(AlpakaCollection const &) = delete;
+  AlpakaCollection &operator=(AlpakaCollection const &) = delete;
+
+  // movable
+  AlpakaCollection(AlpakaCollection &&other) = default;
+  AlpakaCollection &operator=(AlpakaCollection &&other) = default;
+
+  T &operator*() { return layout_; }
+
+  T const &operator*() const { return layout_; }
+
+  T *operator->() { return &layout_; }
+
+  T const *operator->() const { return &layout_; }
+
+  Buffer &buffer() { return *buffer_; }
+
+  Buffer const &buffer() const { return *buffer_; }
+
+private:
+  std::optional<Buffer> buffer_;  //!
+  T layout_;
+};
+
+template <typename T>
+using AlpakaHostCollection = AlpakaCollection<T, alpaka_common::DevHost>;
```

Comment:
Mhm, I see...

The content of the persisted class can be read without any reference to Alpaka (_e.g._ bare ROOT on a laptop).
The full type does remember the dependency on Alpaka though:
```
Type                                  Module                      Label     Process         Full Name
-----------------------------------------------------------------------------------------------------
PortableCollection<XyzIdSoA,alpaka::DevCpu>    "xyzIdAlpakaProducerSerial"   ""        "Writer"        XyzIdSoAalpakaDevCpuPortableCollection_xyzIdAlpakaProducerSerial__Writer
```

@ericcano has been looking for ways around this, playing with inheritance, type-punning, _etc._ but at least some of the possible solutions seem a bit too fragile to me.

I think that - ideally - what we would like to do is to convince ROOT (and the framework) to write only the `T` part of `PortableCollection<T, TDev>`, but allow reading it back into the full `PortableCollection<T, TDev>` object.
But I don't know how we could do that without significantly re-working the branch (name) logic, or the ROOT streaming :man_shrugging:

Reply:
In fact, the ROOT schema evolution should support writing a `T` object and reading it back into a `PortableCollection<T, TDev>` object (basically, this what @ericcano's read streamers already do, since all but the `T` part is only transient).

The problem is the rest :-)
  - how to _write_ an object of type `T` ? maybe this doesn't matter, since `T` is the only persistent part anyway ?
  - how to use a branch name that does not include `alpakaDevCpu` ?
  - how to let the framework read an object of type `T` into an object of type `PortableCollection<T, TDev>` ?
  - _etc._

Maybe we can play tricks with inheritance, or having two separate branches (`T` and `PortableCollection<T, TDev>`) ...